### PR TITLE
Sample function with build and packaging

### PR DESCRIPTION
**Pull request by crissupb**

### CODE REVIEW

File: source/template.yaml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      Policies:
+        - Statement:
+          - Sid: S3ObjectAccess
+            Effect: Allow
+            Action:
+            - s3:PutObject
+            - s3:GetObject
+            - s3:DeleteObject
+            Resource: !Sub
+              - arn:aws:s3:::login-gov-idp-doc-capture-${Environment}.${AccountId}-${Region}/*
```
Comment (zachmargolis) [inline source/template.yaml:75]: Where does this name come from? Our terraform configs?

File: source/template.yaml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+Description: >
+  identity-idp-functions
+
+  IDP functions for login.gov
+
+# More info about Globals: https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst
+Globals:
+  Function:
+    Timeout: 60
+    Runtime: ruby2.7
```
Comment (zachmargolis) [inline source/template.yaml:40]: Ditto, can we make this 2.6?

File: source/demo-function/hello_world/app.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+require 'aws-sdk-s3'
+require 'json'
+require 'logger'
+
+def lambda_handler(event:, context:)
+  # Sample pure Lambda function
```
Comment (zachmargolis) [inline source/demo-function/hello_world/app.rb:6]: can we remove this line? It's not a pure function anymore

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      - run:
+          name: "Package Functions"
+          command: |
+            PATH=/home/linuxbrew/.linuxbrew/bin:$PATH
+            cd source
+            sam package \
+              --profile default \
+              --debug \
+              --s3-bucket $DEPLOY_S3_BUCKET \
+              --s3-prefix circleci/identity-infra-functions/$CIRCLE_BRANCH/packages \
```
Comment (zachmargolis) [inline .circleci/config.yml:95]: do we need this `circleci` here in the bucket path?

File: source/circleci-demo/.gitignore [hunk 5]
```diff
@@ -0,0 +1,244 @@
+
+# Created by https://www.gitignore.io/api/osx,linux,python,windows,pycharm,visualstudiocode
```
Comment (zachmargolis) [inline source/circleci-demo/.gitignore:2]: all the `.gitignore`s in this PR are the same --- could we just use one at the root and not have to worry about them in each directory?

File: source/circleci-demo/hello_world/app.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        Return doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html
+    """
+
+    # try:
+    #     ip = requests.get("http://checkip.amazonaws.com/")
+    # except requests.RequestException as e:
+    #     # Send some context about this error to Lambda Logs
+    #     print(e)
+
+    #     raise e
```
Comment (zachmargolis) [inline source/circleci-demo/hello_world/app.py:36]: can we remove this, if this is the pure function example?

File: source/circleci-demo/hello_world/app.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    return {
+        "statusCode": 200,
+        "body": json.dumps({
+            "message": "hello world",
+            # "location": ip.text.replace("\n", "")
```
Comment (zachmargolis) [inline source/circleci-demo/hello_world/app.py:42]: ditto, remove this?

File: source/template.yaml (hunk 2 above)
Comment (zachmargolis) [reply source/template.yaml:40]: Or are we not able to? Looks like the docs only support 2.5 or 2.7? https://docs.aws.amazon.com/lambda/latest/dg/lambda-ruby.html

File: source/.ruby-version [hunk 8]
```diff
@@ -0,0 +1 @@
+2.7
```
[the remark this answers is not in the document]
Comment (crissupb) [reply source/.ruby-version:1]: Would have to bump back down to 2.5.  2.6 Is not available as a runtime.

File: source/template.yaml (hunk 1 above)
Comment (crissupb) [reply source/template.yaml:75]: Environment is a parameter for the deployment.  AccountId and Region are pulled from the profile being used for deployment.

File: .circleci/config.yml (hunk 4 above)
Comment (crissupb) [reply .circleci/config.yml:95]: Yes.  The bucket permissions given to the circle bot require this folder.

File: source/.ruby-version (hunk 8 above)
[the remark this answers is not in the document]
Comment (zachmargolis) [reply source/.ruby-version:1]: Ok let's keep at 2.7 I guess

**Review by zachmargolis — APPROVED** [review]
LGTM

**Pull request merged**